src/IO/S3/copyDataToS3.cpp

```diff
@@ -0,0 +1,723 @@
+                num_parts = (total_size + part_size - 1) / part_size;
+            }
+
+            if (num_parts < 1 || num_parts > settings.max_part_number || part_size < settings.min_upload_part_size
+                || part_size > settings.max_upload_part_size)
+            {
+                String msg;
+                if (num_parts < 1)
+                    msg = "Number of parts is zero";
+                else if (num_parts > settings.max_part_number)
+                    msg = fmt::format("Number of parts exceeds {}", num_parts, settings.max_part_number);
+                else if (part_size < settings.min_upload_part_size)
+                    msg = fmt::format("Size of a part is less than {}", part_size, settings.min_upload_part_size);
+                else
+                    msg = fmt::format("Size of a part exceeds {}", part_size, settings.max_upload_part_size);
+
+                throw Exception(
+                    ErrorCodes::INVALID_CONFIG_PARAMETER,
+                    "{} while writing {} bytes to S3. Check max_part_number = {}, "
+                    "min_upload_part_size = {}, max_upload_part_size = {}, max_single_part_upload_size = {}",
+                    msg, total_size, settings.max_part_number, settings.min_upload_part_size,
+                    settings.max_upload_part_size, settings.max_single_part_upload_size);
+            }
+
+            /// We've calculated the size of a normal part (the final part can be smaller).
+            normal_part_size = part_size;
+        }
+
+        void uploadPart(size_t part_number, size_t part_offset, size_t part_size)
+        {
+            LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Size: {}", dest_bucket, dest_key, multipart_upload_id, part_size);
+
+            if (!part_size)
+            {
+                LOG_TRACE(log, "Skipping writing an empty part.");
+                return;
+            }
+
+            if (schedule)
+            {
+                UploadPartTask * task = nullptr;
+
+                {
+                    std::lock_guard lock(bg_tasks_mutex);
+                    task = &bg_tasks.emplace_back();
+                    ++num_added_bg_tasks;
+                }
+
+                /// Notify waiting thread when task finished
+                auto task_finish_notify = [this, task]()
+                {
+                    std::lock_guard lock(bg_tasks_mutex);
+                    task->is_finished = true;
+                    ++num_finished_bg_tasks;
+
+                    /// Notification under mutex is important here.
+                    /// Otherwise, WriteBuffer could be destroyed in between
+                    /// Releasing lock and condvar notification.
+                    bg_tasks_condvar.notify_one();
+                };
+
+                try
+                {
+                    task->req = fillUploadPartRequest(part_number, part_offset, part_size);
+
+                    schedule([this, task, task_finish_notify]()
```

Comment:
Help me understand here, how many  concurrent uploads are run here? Is it limited thread pool?
My concern is that it runs a lot of tasks, each makes a new connection. It could run out of available inodes or face other limitation inside CH.

Reply:
It's normally used with `IOThreadPool` . It's limited of course. By default it runs up to 100 threads, so maximum 100 parallel uploads. But it actually must be smaller because `IOThreadPool` is also used for other tasks.

Reply:
In fact it should use its own thread pools `backups_thread_pool` and `restores_thread_pool`, so I changed the code to use the correct thread pools. Those thread pools are created specially for backups and restores.